main/src/ca/uwaterloo/flix/runtime/Solver.scala

```diff
@@ -18,40 +18,90 @@ package ca.uwaterloo.flix.runtime
 
 import ca.uwaterloo.flix.api.{IValue, WrappedValue}
 import ca.uwaterloo.flix.language.ast.ExecutableAst._
+import ca.uwaterloo.flix.language.ast.ExecutableAst.Constraint.Rule
 import ca.uwaterloo.flix.language.ast.{Ast, ExecutableAst, Symbol}
 import ca.uwaterloo.flix.runtime.datastore.DataStore
 import ca.uwaterloo.flix.runtime.debugger.RestServer
-import ca.uwaterloo.flix.util.{Debugger, InternalRuntimeException, Options, Verbosity}
+import ca.uwaterloo.flix.util._
+import java.util.concurrent._
+import java.util.ArrayList
 
 import scala.annotation.tailrec
 import scala.collection.mutable
+import scala.collection.mutable.ArrayBuffer
 
-object Solver {
+/**
+  * Flix Fixed Point Solver.
+  *
+  * Based on a variant of semi-naive evaluation.
+  *
+  * The solver computes the least fixed point of the rules in the given program.
+  */
+class Solver(val root: ExecutableAst.Root, options: Options) {
+
+  //
+  // Types of the solver:
+  //
 
   /**
-    * A case class representing a solver context.
+    * The type of environments.
+    *
+    * An environment is map from identifiers to values.
     */
-  case class SolverContext(root: ExecutableAst.Root, options: Options)
+  type Env = mutable.Map[String, AnyRef]
 
-}
+  /**
+    * The type of work lists.
+    *
+    * A (possibly local) work list is a collection of (rule, env) pairs
+    * specifying that the rule must be re-evaluated under the associated environment.
+    */
+  type WorkList = mutable.ArrayStack[(Rule, Env)]
 
-/**
-  * A solver based on semi-naive evaluation.
-  */
-class Solver(implicit val sCtx: Solver.SolverContext) {
+  /**
+    * The type of a (partial) interpretation.
+    *
+    * An interpretation is collection of facts.
```

Comment:
This comment isn't as clear as the other two above. What does the pair `(Symbol.TableSym, Array[AnyRef])` mean? Is a fact a collection of values (`Array[AnyRef]`) associated with a relation/lattice (`Symbol.TableSym`)?